### PR TITLE
Add ClientConfigXmlGenerator

### PR DESCRIPTION
**Pull request by gurbuzali**

### CODE REVIEW

**Comment by cangencer** [comment]
Can we send PR to Hazelcast as well? I think it can be added in a patch to 3.11?

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/config/ClientConfigXmlGenerator.java [hunk 1]
```diff
@@ -0,0 +1,636 @@
...
+    private static final int INDENT = 5;
+
+    private static final ILogger LOGGER = Logger.getLogger(ConfigXmlGenerator.class);
+
+    private final boolean formatted;
+
+    /**
+     * Creates a new generator instance with {@link #formatted} {@code false}
+     */
+    public ClientConfigXmlGenerator() {
```
Comment (cangencer) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/config/ClientConfigXmlGenerator.java]: looks unused

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/config/ClientConfigXmlGenerator.java [hunk 2]
```diff
@@ -0,0 +1,636 @@
...
+    /**
+     * Transforms the given {@link ClientConfig} to xml string
+     */
+    public String generate(ClientConfig clientConfig) {
+        Preconditions.isNotNull(clientConfig, "ClientConfig");
+
+        StringBuilder xml = new StringBuilder();
+        XmlGenerator gen = new XmlGenerator(xml);
+
+        xml.append("<hazelcast-client ")
```
Comment (cangencer) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/config/ClientConfigXmlGenerator.java]: why not use `.open` here?

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/config/ClientConfigXmlGenerator.java [hunk 3]
```diff
@@ -0,0 +1,636 @@
...
+        }
+
+        private static void appendNode(StringBuilder xml, String name, Object contents, Object... attributes) {
+            if (contents != null || attributes.length > 0) {
+                xml.append('<').append(name);
+                for (int i = 0; i < attributes.length; ) {
+                    Object key = attributes[i++];
+                    Object val = attributes[i++];
+                    if (val != null) {
+                        xml.append(" ").append(key).append("=\"").append(val).append("\"");
```
Comment (cangencer) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/config/ClientConfigXmlGenerator.java]: I don't see any escaping of values here

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/config/ClientConfigXmlGenerator.java [hunk 4]
```diff
@@ -0,0 +1,636 @@
...
+                    }
+                }
+                if (contents != null) {
+                    xml.append('>').append(contents).append("</").append(name).append('>');
```
Comment (cangencer) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/config/ClientConfigXmlGenerator.java]: also here it looks like values are not escaped properly

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/config/ClientConfigXmlGenerator.java [hunk 5]
```diff
@@ -0,0 +1,708 @@
...
+        }
+
+        private static void appendCloseNode(StringBuilder xml, String name) {
+            xml.append("</").append(name).append('>');
+        }
+
+        private static void appendAttributes(StringBuilder xml, Object... attributes) {
+            for (int i = 0; i < attributes.length; ) {
+                xml.append(" ");
+                escapeXml(attributes[i++], xml);
```
Comment (viliam-durina) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/config/ClientConfigXmlGenerator.java]: We don't need to escape attribute names. Rather, we should check them for sanity, I found [these rules](https://stackoverflow.com/a/44057262/952135). I think we can just ignore that, we don't use user-input for attribute names and hard-coded attribute names should be ok.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/config/ClientConfigXmlGenerator.java [hunk 6]
```diff
@@ -0,0 +1,708 @@
...
+                xml.append("=\"");
+                escapeXml(attributes[i++], xml);
+                xml.append("\"");
+            }
+        }
+
+        private static void appendNode(StringBuilder xml, String name, Object contents, Object... attributes) {
+            if (contents != null || attributes.length > 0) {
+                xml.append('<').append(name);
+                for (int i = 0; i < attributes.length; ) {
```
Comment (viliam-durina) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/config/ClientConfigXmlGenerator.java]: We can delegate to `appendAttributes` and other append methods here.

File: hazelcast-jet-core/src/test/java/com/hazelcast/jet/config/ClientConfigXmlGeneratorTest.java [hunk 7]
```diff
@@ -0,0 +1,508 @@
...
+@RunWith(HazelcastParallelClassRunner.class)
+public class ClientConfigXmlGeneratorTest extends JetTestSupport {
+
+    private static final boolean DEBUG = false;
+    private static final Random RANDOM = new Random();
+
+    private ClientConfig clientConfig = new ClientConfig();
+
+    @Test
+    public void escape() {
```
Comment (cangencer) [inline hazelcast-jet-core/src/test/java/com/hazelcast/jet/config/ClientConfigXmlGeneratorTest.java]: missing test for escaping attributes

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/config/ClientConfigXmlGenerator.java [hunk 8]
```diff
@@ -0,0 +1,696 @@
...
+                    case '"':
+                        appendTo.append("&quot;");
+                        break;
+                    case '\'':
+                        appendTo.append("&#39;");
+                        break;
+                    case '&':
+                        appendTo.append("&amp;");
+                        break;
+                    case '<':
```
Comment (cangencer) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/config/ClientConfigXmlGenerator.java]: technically you don't need to escape `<>` in attributes I think

Comment (viliam-durina) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/config/ClientConfigXmlGenerator.java]: You need to escape `<`, you never need to escape `>`

**Pull request merged**